clients/src/main/java/org/apache/kafka/clients/consumer/internals/PartitionAssignor.java

```diff
@@ -58,12 +59,12 @@
      */
     Map<String, Assignment> assign(Cluster metadata, Map<String, Subscription> subscriptions);
 
-
     /**
      * Callback which is invoked when a group member receives its assignment from the leader.
      * @param assignment The local member's assignment as provided by the leader in {@link #assign(Cluster, Map)}
+     * @param generation The consumer group generation associated with this partition assignment (optional)
      */
-    void onAssignment(Assignment assignment);
+    void onAssignment(Assignment assignment, Optional<Integer> generation);
```

Comment:
It's a bit unclear what the compatibility policy should be for plugins, but it seems reasonable to assume that a custom `PartitionAssignor` which does not implement this method will continue to work with the consumer after an upgrade. I think we have to leave the old variant around and add a default implementation which invokes it:
```java
default void onAssignment(Assignment assignment, Optional<Integer> generation) {
  onAssignment(assignment);
}
```
I was debating whether we would want to deprecate the old variant, but it seems not worthwhile since it should be a relatively uncommon case to require the generation.

Reply:
That makes sense. Added in the new commit.